Race condition and test fixes

mvdan opened the pull request:
Picked up from @mikedanese's #12967

The reflect test failure fix really looks like a Go bug. I'll try and find any reference to it.

mvdan committed:
Fix struct input test in jsonpath

mvdan commented:
@mikedanese could you confirm you're okay with this so that `cla: no` can go away?

mikedanese commented:
I'm OK with these contributions. @vishh for cadvisor upgrade, @thockin or @brendandburns for the go upgrade.

mikedanese commented:
@k8s-bot ok to test.

mvdan commented:
`TestHTTPProbeChecker` appears to be failing now, which didn't fail before. Will look into it.

vishh commented:
LGTM for cadvisor changes.

mvdan commented:
Currently talking to a member of the Go team about the weird if condition possible bug.

mvdan commented:
Confirmed - it's a bug in the compiler. My fix works around the issue, so better keep it in to keep the code working in 1.5.

thockin commented:
Achievement Unlocked?

I'm happy to see us support 1.5

On Fri, Aug 21, 2015 at 3:30 PM, Daniel Martí <notifications@github.com>
wrote:

> Confirmed - it's a bug in the compiler. My fix works around the issue, so
> better keep it in to keep the code working in 1.5.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13058#issuecomment-133580001>
> .
>

mvdan commented:
@thockin double achievement unlocked - 1.5 support and compiler bug spotted :)

For anyone interested, the compiler bug is https://github.com/golang/go/issues/12266

mvdan committed:
Fix failing test in probe/http

The error that Go 1.5 returns is different compared to 1.4. Support both.

mvdan commented:
Fixed the `TestHTTPProbeChecker` failure in 1.5, but that will make it fail on 1.4 and earlier. Is that acceptable? Or should we not check GET bodies on `probe.Failure` at all?

smarterclayton commented:
We have to maintain support for building and testing on 1.4 for a while at least.

thockin commented:
of course

On Fri, Aug 21, 2015 at 8:11 PM, Clayton Coleman <notifications@github.com>
wrote:

> We have to maintain support for building and testing on 1.4 for a while at
> least.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13058#issuecomment-133629762>
> .
>

smarterclayton commented:
Was responding to

>  but that will make it fail on 1.4 and earlier. Is that acceptable?

thockin commented:
oh, right.  No, 1.4 has to work.

On Fri, Aug 21, 2015 at 8:15 PM, Clayton Coleman <notifications@github.com>
wrote:

> Was responding to
>
> but that will make it fail on 1.4 and earlier. Is that acceptable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13058#issuecomment-133629935>
> .
>

mvdan commented:
Sure. Amended the last commit to support both the old and new error messages.

Still a few failing tests to go.

mvdan commented:
None of the tests still failing in shippable fail on my machine. Anyone know why isn't shippable running now?

mvdan commented:
Uncovered a bad race condition bug while investigating the portforward test failure on 1.5. It is now fixed, the test passes consistently on my machine now.

I wonder how it could have gone unnoticed in 1.4. Maybe the runtime being slower hid the race condition?

mvdan committed:
Fix probe/tcp test on some systems

Depending on your system, the error might be "Servname not supported for
ai_socktype" instead of "unknown port". Accept both.

For example:
http://www.ducea.com/2006/09/11/error-servname-not-supported-for-ai_socktype/

mvdan commented:
I found another flaky test:

```
--- FAIL: TestRequestAttachRemoteCommand (30.01s)
        remotecommand_test.go:241: 0: expected error stream read 'bail', got 'Timeout occured'
```

This one I can't reproduce, so I'm ignoring it for now. Or should I open an issue about it?

mvdan commented:
In fact, shippable had errors in remotecommand too. Although the errors are different, it probably means that there's another underlying bug that 1.5 is uncovering. Will investigate.

mvdan commented:
Fixed too. Let's see if shippable passes now.

Any chance we could have shippable run five times in a row, to be sure there aren't any more race condition bugs popping up thanks to 1.5?

mvdan commented:
All the tests are passing, but `hack/test-cmd.sh` is failing due to a template issue.

Seems like the template package used to allow `{{..foo}}` instead of `{{.foo}}`, and 1.5 simply doesn't. Will try a fix.

mvdan committed:
Avoid using two periods in templates

Since $id_field already starts with a period, .$id_field would result in the
following in Go 1.5:

error: error parsing template {{range.items}}{{..metadata.name}}:{{end}}, template: output:1: unexpected . after term "."

Apparently, Go 1.4 allowed this. Unnecessary anyway.

mvdan commented:
Success! There's only one integration test failing now. Will look at it tomorrow.

mvdan committed:
Fix race condition in watch

The Shutdown() call returned immediately, without waiting for all event
distributions to be completed. Even worse, it would close all the watcher
result channels before all the info was sent to them.

Properly wait for all distributor goroutines - currently only one - to be
finished.

This fixes the flaky test TestBroadcasterDropIfChannelFull. Bonus cleanup on
said test too.

mvdan commented:
Shippable uncovers more and more flaky tests due to race conditions at every run, so don't know when I'll be done. Merging in 1.5 before these are fixed would be unwise, since apparently they aren't flaky when running on 1.4.

mvdan commented:
It just occured to me that all the new flaky tests are showing up because the scheduler in 1.4 is deterministic, but the one in 1.5 is randomized.

All the bugs I'm fixing now aren't present in 1.4, because we were lucky that the deterministic scheduling didn't trigger the race conditions.

mvdan committed:
Try to fix flaky test in record/event

The error messages were inconsistent with what was actually being tested in
regards to timestamps being equal or not.

mvdan commented:
Shippable and jenkins have passed three times in a row, so I would say this is ready.

mvdan commented:
Any reviewers?

mvdan commented:
Rebased on master, tests still pass on my machine.

mvdan commented:
Oh, go's `-race` found another bug. Will fix.

liggitt commented on pkg/client/unversioned/remotecommand/remotecommand_test.go:
```diff
@@ -92,15 +92,21 @@ func fakeExecServer(t *testing.T, i int, stdinData, stdoutData, stderrData, erro
 		if len(stderrData) > 0 {
 			fmt.Fprint(stderrStream, stderrData)
+		}
+		if stderrStream != nil {
```
I touched this in https://github.com/kubernetes/kubernetes/pull/13144

It's actually better to avoid calling Close() on stdoutStream/stderrStream and let the deferred Reset() calls close the streams. Calling Close() explicitly actually introduces a race with the stdinStream read below

mvdan replied:
If I remove the Close() calls, the tests just hang. The commit message explains why - if we don't close the error stream, another goroutine blocks forever waiting for more errors.

Does your PR fix that?

liggitt replied:
remove Close() and switch the stdin ioutil.ReadFull to io.ReadAll(expectedLen)

liggitt replied:
This test shouldn't have to close anything explicitly... all the streams get Reset() and conn.Close() is called when the function returns. The ioutil.ReadFull() below was what was hanging

mvdan replied:
I see, trying that out now.

liggitt commented on pkg/client/unversioned/remotecommand/remotecommand.go:
```diff
@@ -155,18 +156,16 @@ func (e *Streamer) doStream() error {
+	var wg sync.WaitGroup
 
 	cp := func(s string, dst io.Writer, src io.Reader) {
+		defer wg.Done()
```
this gets called for stdin, which won't terminate until we tear it down

mvdan replied:
I did not run into that stdin issue with the unit tests, see the two commits in this PR touching this file.

Or does that problem show up somewhere other than unit/integration tests?

liggitt replied:
The unit tests mock reality... @ncdc can probably give more details about the behavior of stdin streams in real life

mvdan replied:
I just knew that not waiting for stdin to finish reading is definitely wrong, so it seemed to me to better try and trigger the bug than to leave in code that I knew was wrong. And I just failed at reproducing the bug, so I left that in :)

liggitt commented on pkg/client/unversioned/remotecommand/remotecommand.go:
```diff
@@ -195,49 +195,37 @@ func (e *Streamer) doStream() error {
-		case err := <-errorChan:
-			return err
-		}
+	wg.Wait()
```
waiting here means a hung stdout/stderr stream can keep us from returning, even if we encountered an error

mvdan replied:
That was the original intended behaviour. What do you suggest instead of waiting?

liggitt replied:
more noting the change than objecting to it... I like waiting for all three streams, but I wasn't sure if there was a situation where an error propagated via the errorStream was expected to exit early and save us from a hung stderr/stdout stream

liggitt commented on pkg/client/unversioned/remotecommand/remotecommand.go:
```diff
@@ -155,18 +156,16 @@ func (e *Streamer) doStream() error {
 
-	doneChan := make(chan struct{}, 2)
 	errorChan := make(chan error)
+	var wg sync.WaitGroup
```
cc @ncdc

liggitt commented on pkg/client/unversioned/remotecommand/remotecommand.go:
```diff
@@ -195,49 +195,37 @@ func (e *Streamer) doStream() error {
-		// because stdin is not closed until the process exits. If we try to call
-		// stdin.Close(), it returns no error but doesn't unblock the copy. It will
-		// exit when the process exits, instead.
+		wg.Add(1)
```
@ncdc will waiting stdin to close itself hang here?

mvdan commented:
Another race condition fixed.

There's an e2e test that keeps on failing, anyone got any ideas? This should speak to someone:

`Unexpected number of endpoints: found map[92ef79f7-4b58-11e5-90f1-42010af0bcc6:[100]], expected map[podname1:[100] podname2:[101]]`

Part of this test:

```
[It] should serve multiport endpoints from pods
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:236
```

mvdan commented:
@liggitt with the stdin changes reverted, remotecommands is flaky again. I'll keep on digging...

liggitt commented:
Do you have a copy of the flake?

mvdan commented:
@liggitt here you go:

```
--- FAIL: TestRequestExecuteRemoteCommand (0.01s)
        remotecommand_test.go:184: 1: expected stderr data 'c', got ''
```

Although, surprisingly enough, making it wait for stdin to finish copying doesn't fix it. Currently running the test 100 times to check for flakyness.

liggitt commented:
Yeah, the readfull prior to closing any of the streams that execute is waiting for should have covered it :-/

mvdan commented:
It doesn't appear to be a race condition - even a `time.Sleep(100 * time.Millisecond)` before the `wg.Wait()` doesn't fix it.

mvdan commented:
@liggitt found it - it's the fake exec server that is racey. Making it only handle one HTTP connection at a time via a mutex fixes the issue. This one was hard to find...

mvdan commented:
The race detected in the oidc package isn't actually our fault: https://github.com/golang/go/issues/12262

mvdan committed:
Fix race condition in controller_test

Our WaitGroup.Add() call might happen after some WaitGroup.Done() calls done
by the controller, so make sure that doesn't happen by doing the Add() calls
before letting the controller run.

--- FAIL: TestUpdate (2.00s)
panic: sync: WaitGroup is reused before previous Wait has returned [recovered]
        panic: sync: WaitGroup is reused before previous Wait has returned

mvdan commented:
To summarize:

* There is a flake in remotecommand (see comments above) but it's not introduced by Go 1.5, as @liggitt mentioned it happens on 1.4 too
* There is a flaky e2e test: `[It] should serve multiport endpoints from pods` repeatedly fails with `Unexpected number of endpoints` until timeout is reached. Has anyone seen it before? Any ideas?

smarterclayton commented:
Services e2e tests tend to be flaky because they are dependent on
scheduling several pods and wait a short time period while parallel tests
are running.  I don't believe we have a real flake there, but we haven't
proved it yet.

On Aug 26, 2015, at 3:11 AM, Daniel Martí <notifications@github.com> wrote:

To summarize:

   - There is a flake in remotecommand (see comments above) but it's not
   introduced by Go 1.5, as @liggitt <https://github.com/liggitt> mentioned
   it happens on 1.4 too
   - There is a flaky e2e test: [It] should serve multiport endpoints from
   pods repeatedly fails with Unexpected number of endpoints until timeout
   is reached. Has anyone seen it before? Any ideas?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13058#issuecomment-134850249>
.

mvdan commented:
So since as far as I can tell this PR doesn't add any flakyness, I think it's ready to be merged.

Ping @mikedanese @thockin

thockin commented:
@mikedanese for final

ncdc commented:
I just noticed there are port forwarding changes. I need some time to review as I'm afraid they might cause leaked goroutines.

ncdc commented:
I have a PR that I'm working on to address some port forward issues (in relation to #8766) but I'm spread a bit thin at the moment and haven't had time to devote to it recently.

ncdc commented:
Actually, if the tests pass, this works for me. I'm working on some new e2e tests that will test various port forwarding scenarios to make sure there aren't any hangs or leaks.

mvdan commented:
@ncdc thanks for the info. I only did the minimal to fix the race conditions and get the tests to consistently pass - I highly doubt that introduced any bugs, especially in portforward where the changes are minimal.

ncdc commented:
@mvdan my fear is this won't ever be called due to the root cause of 8766: https://github.com/kubernetes/kubernetes/pull/13058/files#diff-3c23f834eaf7e756813f19c5a1116c46R295

mvdan commented:
@ncdc so you're saying that currently two bugs cancel each other out, and that if I fix this one the other will make the `Wait()` call to block forever?

mikedanese commented:
@mvdan thanks for fixing all these races. I will try to review shortly. However I am hesitant to merge the go1.5 change. I dislike that we had to restructure correct if statements to make go1.5 work. Doesn't it seem like we should wait for go1.5.1 to fix some of the multiple regressions we ran into here before bumping the version we run in CI?

mvdan commented:
@mikedanese following the same logic, why are we using 1.4 instead of 1.4.2?

1.5.1 is scheduled for mid-September. Waiting for that to bump the CI version seems fine to me - I can remove your commit from this PR - but the rest of commits should still be merged.

@ncdc and myself have come to a solution regarding portforward, I'll amend the commit in a second.

liggitt commented:
+1 on merging race fixes

mikedanese commented:
We didn't have to restructure correct if statements when we moved to 1.4 so the same logic does not apply. I agree that we should merge race fixes. I only said "I am hesitant to merge the go1.5 changes".

mvdan commented:
@mikedanese little bugs like these got fixed between 1.4 and 1.4.2, surely some affect kubernetes and weren't caught in unit/e2e tests. I agree that this bug is probably bigger than any other, but I would always use the latest stable release of each major Go release regardless of whether it fixes any regressions/bugs or not.

Just to confirm - by go1.5 changes, you mean the shippable changes and any changes that work around bugs in 1.5, right?

mvdan commented:
In fact, the only go 1.5 workaround there is the if change, which doesn't break 1.4. If we don't merge it, kubernetes will remain broken on the current Go release until 1.5.1 is out. I think that's worse.

mvdan commented:
I have removed the portforward race fix since it requires server side fixes too, and it isn't introduced by 1.5. @ncdc is taking care of it in https://github.com/kubernetes/kubernetes/pull/12283.

mvdan commented:
Also removed the remotecommand race fix since it's been cherry picked by @liggitt in #13144

timothysc commented:
@wojtek-t @fgrzadkowski very interested to see the gc-perf difference on 1.5 for the api-server.

mikedanese commented:
@mvdan I am referring to only the shippable.yml and .travis.yml changes when I refer to go1.5 changes. All the bug fixes are definitely worthwhile. I will review this tomorrow or this weekend. If other people feel strongly that we should bump ci to go1.5 I won't veto, but  in light of the nasty if bug I tend against it.

mvdan commented:
@mikedanese agreed. I'll remove your commit from this PR.

mikedanese commented on third_party/forked/reflect/deep_equal.go:
```diff
@@ -156,7 +156,9 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
 		return true
 	case reflect.Slice:
-		if (v1.IsNil() || v1.Len() == 0) != (v2.IsNil() || v2.Len() == 0) {
+		empty1 := v1.IsNil() || v1.Len() == 0
```
can you leave a comment here that explains why this change was necessary?

mvdan replied:
sure.

mikedanese commented on pkg/probe/http/http_test.go:
```diff
@@ -39,14 +48,31 @@ func TestHTTPProbeChecker(t *testing.T) {
-		body    string
+		handler   func(w http.ResponseWriter)
+		health    probe.Result
+		accBodies []string
```
can you leave a comment here that explains why this change was necessary? (potentially include something I could grep for like `// go1.5: the error message changed`)

mvdan replied:
sure.

mikedanese commented on pkg/client/unversioned/record/event_test.go:
```diff
@@ -316,17 +319,17 @@ func validateEvent(actualEvent *api.Event, expectedEvent *api.Event, t *testing.
 		t.Errorf("timestamp wasn't set: %#v", *actualEvent)
 	}
-	if actualEvent.FirstTimestamp.Equal(actualEvent.LastTimestamp) {
+	actualFirstTimestamp := actualEvent.FirstTimestamp
```
go1.5 comment please

mvdan replied:
This isn't a go1.5 workaround/fix. Just a small code simplification I did along the way while I fixed the race bug.

mikedanese replied:
Ah ok.

mikedanese commented on pkg/client/unversioned/record/event_test.go:
```diff
@@ -286,24 +289,24 @@ func TestEventf(t *testing.T) {
 		eventBroadcaster := NewBroadcaster()
 		sinkWatcher := eventBroadcaster.StartRecordingToSink(&testEvents)
 		logWatcher1 := eventBroadcaster.StartLogging(t.Logf) // Prove that it is useful
+		wg.Add(1)
```
Instead of adding a wait group, why not just move one of the `<-called` to this line? Isn't that the same?

mvdan replied:
A waitgroup is simpler to use and better (e.g. it lets goroutines die sooner).

The original code did the waiting right before the Stop() calls, so I did the same with a waitgroup. Splitting the waiting in two separate blocks of code would confuse things, especially since you cannot know which of the two channel receives would correspond to which of the channel sends.

lavalamp replied:
Why not start the wait group off with 2 to begin with?

mvdan replied:
The idea is to keep each `Add(1)` close to what will later trigger the `Done()` call. If you do a single `Add(n)` at the top, it might later be harder to spot what blocks of code amount for how many of the `Done()` calls.

This of course doesn't apply to loops.

liggitt replied:
+1 for pairing Add/Done calls

mikedanese commented:
https://github.com/mvdan/kubernetes/commit/2527da99e1f826febc0cd97b0d3b42a1713f08b6 looks fine to me but cc @karlkfi @sttts @jdef for a second review on that commit.

mikedanese commented:
Generally LGTM. I would like @lavalamp to review https://github.com/mvdan/kubernetes/commit/579118752230211aca5eb23e74dc4ccfc6457590  before final approval.

mvdan commented:
FWIW, after the rebase the two commits that @mikedanese linked to are now https://github.com/mvdan/kubernetes/commit/d3490e86ec34abd785d0a1974b09738149ce71ef and https://github.com/mvdan/kubernetes/commit/2ab6033cb0b94c3321796893e34744fedc9a9a46, respectively.

lavalamp commented on third_party/forked/reflect/deep_equal.go:
```diff
@@ -156,7 +156,10 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
 		return true
 	case reflect.Slice:
-		if (v1.IsNil() || v1.Len() == 0) != (v2.IsNil() || v2.Len() == 0) {
+		// go1.5: simplified if to work around https://github.com/golang/go/issues/12266
```
Let's not have these workarounds in our code, please--let's wait for a go1.5 patch before using go1.5

mvdan replied:
You can argue that they should be removed once the current version of Go is 1.5.1. But for as long as the stable is 1.5, not including this patch will mean that you can't use Kubernetes with the current stable version, not even for testing/development purposes.

lavalamp commented on pkg/watch/mux.go:
```diff
@@ -149,6 +151,7 @@ func (m *Broadcaster) Action(action EventType, obj runtime.Object) {
 // You must not call Action after calling Shutdown.
```
Action or Watch

lavalamp replied:
Also, say that it blocks until all events have been distributed to all extant watchers.

mvdan replied:
Sure.

lavalamp replied:
Clarification: distributed to, but not necessarily handled by, all extant watchers; stuff could remain sitting in the outbound channels.

mvdan replied:
You mean blocked channel sends because the receivers aren't listening yet? In that case, since our channel send has blocked waiting for a listener, the Done() call would not be done and we would wait. So all events must have been sent, nothing can remain sitting in outbound channels - at least with the current approach.

lavalamp replied:
Outbound channels could be buffered, and even if they weren't, you don't wait for close to have been called on them.

mvdan replied:
I had not noticed that our channels could be buffered, you're right.

I can't find any reference to this, but I am guessing that if you close a buffered channel, it can still be read from and its buffered values are not lost.

lavalamp replied:
>  if you close a buffered channel, it can still be read from and its buffered values are not lost.

Yes, that's correct.

lavalamp commented on pkg/watch/mux.go:
```diff
@@ -162,6 +165,7 @@ func (m *Broadcaster) loop() {
 		}
 		m.distribute(event)
 	}
+	m.distributing.Done()
```
Do this after the closeAll()?

mvdan replied:
Which would result in closing the channels before we waited for all the info to have been sent to them. I don't see why.

lavalamp replied:
No-- a channel close doesn't blow away all the data in the channel, it just prevents further writes.

mvdan replied:
The writes may not even have started, that's my point.

lavalamp replied:
Distribute is called synchronously above. Once we get to this point, no more writes to the outgoing channels will occur, unless you count the close call, which I do. After closeAll is called, then there really will not be any more writes to the channels and it should be safe to tell the waitgroup that we're done.

mvdan replied:
You are right, the channel sends are synchronous and cannot happen at this point. And they must all have been received by this point too, so even if closing a channel cancels blocked unbuffered channel sends - which I'm not sure about - this would never happen.

Nice catch!

lavalamp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -371,11 +366,18 @@ func TestUpdate(t *testing.T) {
 
-	// run every test a few times, in parallel
 	const threads = 3
+	testDoneWG.Add(threads * len(tests))
```
Nit: slight preference for moving this line up instead of moving the controller start down here.

mvdan replied:
Sure. My intention was to keep `const threads = 3` close to the code that uses it below, but what you suggest would mean fewer line changes so it makes sense.

lavalamp commented on pkg/client/unversioned/record/event_test.go:
```diff
@@ -271,14 +272,16 @@ func TestEventf(t *testing.T) {
 					t.Errorf("Expected event update(), got event create()")
 				}
-				called <- struct{}{}
+				wg.Done()
```
Use `defer wg.Done()` at the beginning of the callback.

mvdan replied:
Nice catch. Although, to my defense, this error was already there :)

jdef commented on contrib/mesos/pkg/proc/proc.go:
```diff
@@ -122,11 +121,13 @@ func (self *procImpl) begin() runtime.Signal {
 	var entered runtime.Latch
 
 	// execute actions on the backlog chan
+	self.wg.Add(1)
```
this package has been completely re-implemented, see #10457 (lgtm'd, awaiting merge)

can we avoid making changes here for now?

mvdan replied:
Sure.

mvdan commented:
Thanks for the throrough code review @lavalamp! Much appreciated.

lavalamp commented:
Sure, no problem, thanks for clean up!

mvdan commented:
All comments amended.

mikedanese commented:
@k8s-bot ok to test

gmarek commented:
cc @gmarek @fgrzadkowski @wojtek-t

mvdan commented:
Go 1.5.1 was just released: https://github.com/golang/go/issues?q=milestone%3AGo1.5.1

@mikedanese enabling that in the CI could now be done.

mikedanese commented:
@mvdan please do reenable. LGTM when CI is green

mvdan commented:
`gimme` doesn't know about 1.5.1 yet, it was released just hours ago. I'll look again tomorrow.

mikedanese commented:
@mvdan sounds good. for context, I was trying to switch shippable to gimme, because IIUC gvm builds go from source and post 1.5 building from source requires a compiled go to bootstrap the compiler. Thus we would have to build multiple versions of go per shippable run

timothysc commented:
/cc @ingvagabund  fyi.

mvdan commented:
@mikedanese in fact, let's keep the 1.5.1 bump separate. You might want to build kubernetes with 1.5.1 instead of 1.4 for official builds too.

Shippable failed, but I keep on getting 500's on all of its builds. This PR should not be adding any failures anyway. If it does, or if there are any more issues arising thanks to 1.5, feel free to ping me.

mikedanese commented:
Shippable says all tests pass but
```
+++ [0909 17:39:00] Saved JUnit XML test report to /root/workspace/src/github.com/kubernetes/kubernetes/shippable/testresults/junit_echo_20150909-172222.xml
!!! Error in ./hack/test-go.sh:210
  'return ${rc}' exited with status 2
Call stack:
  1: ./hack/test-go.sh:210 main(...)
Exiting with status 1
```
@ixdy do you know what's going on here? It's not readily obvious.

mvdan commented:
@mikedanese shippable was acting up yesterday, maybe it will work properly now.

Removed the Go 1.5 bug workaround, since 1.5.1 is now out.

mikedanese commented:
Fixes #13764

ixdy commented on pkg/controller/framework/controller_test.go:
```diff
@@ -335,7 +335,12 @@ func TestUpdate(t *testing.T) {
 	)
 
+	const threads = 3
+	testDoneWG.Add(threads * len(tests))
```
The issue with Shippable is that this test is failing to build, probably because `tests` isn't defined yet when it's used here.

mvdan replied:
You're right. Whoops.

This used to compile, but when @lavalamp suggested to reorganize the code I didn't see this error. The only sane way to fix this is to move the code back where it was, effectively undoing the suggestion from Daniel.

lavalamp replied:
Oops.
